ddl/ddl_api.go

```diff
@@ -6298,6 +6300,126 @@ func (d *ddl) AlterTablePartitionAttributes(ctx sessionctx.Context, ident ast.Id
+	}
+
+	meta := tb.Meta()
+	if meta.Partition == nil {
+		return errors.Trace(ErrPartitionMgmtOnNonpartitioned)
+	}
+
+	partitionID, err := tables.FindPartitionByName(meta, spec.PartitionNames[0].L)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	ptDef := &model.PartitionDefinition{}
+	if spec.Options != nil {
+		for _, op := range spec.Options {
+			switch op.Tp {
+			case ast.TableOptionPlacementPolicy:
+				ptDef.PlacementPolicyRef = &model.PolicyRefInfo{
+					Name: model.NewCIStr(op.StrValue),
+				}
+			case ast.TableOptionPlacementPrimaryRegion:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.PrimaryRegion = op.StrValue
+			case ast.TableOptionPlacementRegions:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Regions = op.StrValue
+			case ast.TableOptionPlacementFollowerCount:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Followers = op.UintValue
+			case ast.TableOptionPlacementVoterCount:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Voters = op.UintValue
+			case ast.TableOptionPlacementLearnerCount:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Learners = op.UintValue
+			case ast.TableOptionPlacementSchedule:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Schedule = op.StrValue
+			case ast.TableOptionPlacementConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.Constraints = op.StrValue
+			case ast.TableOptionPlacementLeaderConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.LeaderConstraints = op.StrValue
+			case ast.TableOptionPlacementLearnerConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.LearnerConstraints = op.StrValue
+			case ast.TableOptionPlacementFollowerConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.FollowerConstraints = op.StrValue
+			case ast.TableOptionPlacementVoterConstraints:
+				if ptDef.DirectPlacementOpts == nil {
+					ptDef.DirectPlacementOpts = &model.PlacementSettings{}
+				}
+				ptDef.DirectPlacementOpts.VoterConstraints = op.StrValue
+			default:
+				return errors.Trace(errors.New("unknown placement policy option"))
```

Comment:
It is table option, not only placement options.

Reply:
done